sermant-agentcore/sermant-agentcore-implement/src/main/java/io/sermant/implement/service/xds/discovery/XdsServiceDiscoveryImpl.java

```diff
@@ -93,6 +91,47 @@ public Set<ServiceInstance> getServiceInstance(String serviceName) {
         return XdsDataCache.getServiceInstance(serviceName);
     }
 
+    /**
+     * get service instance of service cluster
+     *
+     * @param clusterName cluster name
+     * @return XdsClusterInstance
+     */
+    @Override
+    public Optional<XdsClusterLoadAssigment> getClusterServiceInstance(String clusterName) {
+        CountDownLatch countDownLatch = new CountDownLatch(1);
+
+        Optional<String> serviceNameOptional = XdsCommonUtils.getServiceNameFromCluster(clusterName);
+        if (!serviceNameOptional.isPresent()) {
+            return Optional.empty();
+        }
+
+        String serviceName = serviceNameOptional.get();
+
+        // first check the cache and return if service instance exists
+        if (XdsDataCache.isContainsRequestObserver(serviceName)) {
+            return XdsDataCache.getClusterServiceInstance(serviceName, clusterName);
+        }
+
+        // locking ensures that a service only creates one stream
+        LOCK.lock();
+        try {
+            // check the cache again after locking and return if service instance exists
+            if (XdsDataCache.isContainsRequestObserver(serviceName)) {
+                return XdsDataCache.getClusterServiceInstance(serviceName, clusterName);
+            }
```

Comment:
You are advised to create CountDownLatch only after all judgments are passed.

Reply:
done